File: src/main/csharp/Session.cs
using System.Runtime.InteropServices;
using Microsoft.Extensions.Logging;
using Structmap.WebTransportFast.Native;

namespace Structmap.WebTransport;

public unsafe record struct Session(Server Server, Object Identifier)
{
    public record Start(Session Session);
    public record End(Session Session);

    public ILogger Logger => Server.Logger;

    /// <summary>
    /// Initiate session close by invoking wtf_session_close
    /// </summary>
    /// <param name="session">Session</param>
    /// <param name="errorCode">No error 0x00, internal error 0x01, etc</param>
    /// <param name="reason">Short message to help when troubleshooting</param>
    /// <returns>Boolean success or failure</returns>
    public bool Close(int errorCode = 0, string reason = "")
    {
        var ptr = (IntPtr)Identifier;
        var n = reason.Length;
        sbyte* reasonBytes = stackalloc sbyte[n + 1];
        for (int i = 0; i < n; i++)
        {
            reasonBytes[i] = (sbyte)reason[i];
        }
        reasonBytes[n] = (sbyte)'\0';
        var result = Methods.wtf_session_close((wtf_session*)ptr, (uint) errorCode, reasonBytes);
        if (result != wtf_result_t.WTF_SUCCESS)
        {
            var msg = Marshal.PtrToStringAnsi((IntPtr)Methods.wtf_result_to_string(result));
            Server.Logger.LogDebug("Failed to close session: 0x{session:x}", ptr);
            return false;
        }
        Logger.LogDebug("Closing session: 0x{session:x}", ptr);
        return true;
    }
    /// <summary>
    /// Initiate session drain by invoking wtf_session_drain
    /// </summary>
    /// <param name="session">Session</param>
    /// <returns>Boolean success or failure</returns>
    public bool Drain()
    {
        var ptr = (IntPtr)Identifier;
        var result = Methods.wtf_session_drain((wtf_session*)ptr);
        if (result != wtf_result_t.WTF_SUCCESS)
        {
            var msg = Marshal.PtrToStringAnsi((IntPtr)Methods.wtf_result_to_string(result));
            Logger.LogDebug("Failed to drain session: 0x{session:x}", ptr);
            return false;
        }
        Logger.LogDebug("Draining session: 0x{session:x}", ptr);
        return true;
    }
    /// <summary>
    /// Initiate stream by invoking wtf_session_create_stream
    /// </summary>
    /// <param name="session">Session on which to initiate</param>
    /// <param name="input">Data to send to client</param>
    /// <param name="bidirectional">Boolean to allow client reply or not</param>
    /// <returns>Stream for client's reply when bidirectional, or Stream.Null if unidirectional.</returns>
    public System.IO.Stream Push(Session session, System.IO.Stream input, bool bidirectional = true)
    {
        throw new NotImplementedException();
    }
    /// <summary>
    /// Check if session is in handshaking state, likely to be ephemeral
    /// </summary>
    /// <param name="session">Session</param>
    /// <returns>True if WTF_SESSION_HANDSHAKING, false otherwise</returns>
    public bool IsHandshaking()
    {
        return wtf_session_state_t.WTF_SESSION_HANDSHAKING == Methods.wtf_session_get_state((wtf_session*)(IntPtr)Identifier);
    }
    /// <summary>
    /// Check if session was recently in connected state
    /// </summary>
    /// <param name="session">Session</param>
    /// <returns>True if WTF_SESSION_CONNECTED, false otherwise</returns>
    public bool IsConnected()
    {
        return wtf_session_state_t.WTF_SESSION_CONNECTED == Methods.wtf_session_get_state((wtf_session*)(IntPtr)Identifier);
    }
    /// <summary>
    /// Check if session is draining
    /// </summary>
    /// <param name="session">Session</param>
    /// <returns>True if WTF_SESSION_DRAINING, false otherwise</returns>
    public bool IsDraining()
    {
        return wtf_session_state_t.WTF_SESSION_DRAINING == Methods.wtf_session_get_state((wtf_session*)(IntPtr)Identifier);
    }
    /// <summary>
    /// Check if session is closed
    /// </summary>
    /// <param name="session">Session</param>
    /// <returns>True if WTF_SESSION_CLOSED, false otherwise</returns>
    public bool IsClosed()
    {
        return wtf_session_state_t.WTF_SESSION_CLOSED == Methods.wtf_session_get_state((wtf_session*)(IntPtr)Identifier);
    }
}

File: src/main/java/com/structmap/webtransport/Session.java
package com.structmap.webtransport;

import com.structmap.webtransportfast.wtf_h;

import java.lang.foreign.Arena;
import java.lang.foreign.MemorySegment;

import static com.structmap.webtransport.Server.logger;

public record Session(Server server, Object identifier) {

    public record Start(Session session) {
    }

    public record End(Session session) {
    }

    /**
     * @param errorCode No error 0x00, internal error 0x01, etc
     * @param reason    Short message for troubleshooting
     * @return true unless there was an internal error
     */
    public boolean close(int errorCode, String reason) {
        var ptr = (MemorySegment)this.identifier;

        try (Arena arena = Arena.ofConfined()) {
            MemorySegment reasonPtr = arena.allocateFrom(reason);
            int result = wtf_h.wtf_session_close(ptr, errorCode, reasonPtr);
            if (result != wtf_h.WTF_SUCCESS()) {
                var msg = wtf_h.wtf_result_to_string(result);
                logger.warn("Failed to close session 0x{}: {}",
                        Long.toHexString(ptr.address()), msg.getString(0));
                return false;
            }
            return true;
        }
    }

    /**
     * @return true unless there was an internal error
     */
    public boolean drain() {
        var ptr = (MemorySegment)this.identifier;
        int result = wtf_h.wtf_session_drain(ptr);
        if (result != wtf_h.WTF_SUCCESS()) {
            var msg = wtf_h.wtf_result_to_string(result);
            logger.warn("Failed to drain session 0x{}: {}",
                    Long.toHexString(ptr.address()), msg.getString(0));
            return false;
        }
        return true;
    }

    /**
     * @return true if session is in handshaking state
     */
    public boolean isHandshaking() {
        var ptr = (MemorySegment)this.identifier;
        return wtf_h.WTF_SESSION_HANDSHAKING() == wtf_h.wtf_session_get_state(ptr);
    }

    /**
     * @return true if session is connected
     */
    public boolean isConnected() {
        var ptr = (MemorySegment)this.identifier;
        return wtf_h.WTF_SESSION_CONNECTED() == wtf_h.wtf_session_get_state(ptr);
    }

    /**
     * @return true if session is draining
     */
    public boolean isDraining() {
        var ptr = (MemorySegment)this.identifier;
        return wtf_h.WTF_SESSION_DRAINING() == wtf_h.wtf_session_get_state(ptr);
    }

    /**
     * @return true if session is closed
     */
    public boolean isClosed() {
        var ptr = (MemorySegment)this.identifier;
        return wtf_h.WTF_SESSION_CLOSED() == wtf_h.wtf_session_get_state(ptr);
    }
}
